### PR TITLE
Uses map data structure to make search fast

### PR DESCRIPTION
**Pull request by tremblerz**

### CODE REVIEW

**Comment by dmiller-nmap** [comment]
I ran into a segfault when running this against 4096 random IP addresses.

```
mass_rdns: 2.66s 2400/4096 [#: 3, OK: 884, NX: 1511, DR: 5, SF: 216, TR: 2765]

Program received signal SIGSEGV, Segmentation fault.
DNS::Factory::ipToPtr (ip=..., ptr="") at nmap_dns.cc:1318
1318      switch (ip.ss_family) {
(gdb) bt
#0  DNS::Factory::ipToPtr (ip=..., ptr="") at nmap_dns.cc:1318
#1  0x000000000044d00f in DNS::Factory::buildReverseRequest (ip=..., buf=buf@entry=0x7fffffffbc40 " ", maxlen=maxlen@entry=512) at nmap_dns.cc:1436
#2  0x000000000044de03 in put_dns_packet_on_wire (req=req@entry=0xef04b0) at nmap_dns.cc:541
#3  0x000000000044e03f in do_possible_writes () at nmap_dns.cc:510
#4  0x00000000004f6138 in event_dispatch_and_delete (nsp=nsp@entry=0xf2a7d0, nse=nse@entry=0xf2b640, notify=notify@entry=1) at nsock_event.c:373
#5  0x00000000004f351c in process_event (nsp=nsp@entry=0xf2a7d0, evlist=evlist@entry=0xf2a818, nse=nse@entry=0xf2b640, ev=ev@entry=3) at nsock_core.c:1075
#6  0x00000000004f39a5 in process_iod_events (nsp=nsp@entry=0xf2a7d0, nsi=nsi@entry=0xf2b330, ev=3) at nsock_core.c:1140
#7  0x00000000004f8c49 in iterate_through_event_lists (evcount=<optimized out>, nsp=0xf2a7d0) at engine_epoll.c:347
#8  epoll_loop (nsp=0xf2a7d0, msec_timeout=<optimized out>) at engine_epoll.c:314
#9  0x00000000004f3373 in nsock_engine_loop (msec_timeout=61, nsp=0xf2a7d0) at nsock_internal.h:414
#10 nsock_loop (nsp=0xf2a7d0, msec_timeout=96) at nsock_core.c:942
#11 0x000000000044ecb4 in nmap_mass_rdns_core (targets=targets@entry=0x941c60, num_targets=num_targets@entry=4096) at nmap_dns.cc:1170
#12 0x000000000044ff86 in nmap_mass_rdns (targets=0x941c60, num_targets=4096) at nmap_dns.cc:1270
#13 0x000000000049fb6d in refresh_hostbatch (hs=hs@entry=0x7fffffffd130, exclude_group=<optimized out>, ports=0x91a640 <ports>, pingtype=1) at targets.cc:730
#14 0x00000000004a0038 in nexthost (hs=hs@entry=0x7fffffffd130, exclude_group=exclude_group@entry=0x7fffffffd090, ports=ports@entry=0x91a640 <ports>, pingtype=<optimized out>)
    at targets.cc:736
#15 0x000000000045a6ef in nmap_main (argc=argc@entry=7, argv=argv@entry=0x7fffffffddd8) at nmap.cc:1984
#16 0x0000000000430099 in main (argc=7, argv=0x7fffffffddd8) at main.cc:228
```

I'm not sure if this is related, but it seems to me like adding the request to the `records` map should be done in `write_evt_handler` instead of `put_dns_packet_on_wire`, since that's also where we add it to the `in_process` list. 

Also, we should verify in `process_result` that the IP in the response is the same one we were asking for. This is the removed line that should be restored in some way:

    if(!result.empty() && (!sockaddr_storage_equal(&ip, tpreq->targ->TargetSockAddr())))
      continue;

File: nmap_dns.cc [hunk 1]
```diff
@@ -599,7 +616,11 @@ static int deal_with_timedout_reads() {
             output_summary();
             stat_dropped++;
             total_reqs--;
-            delete tpreq;
+            infoI = records.find(tpreq->id);
+            if ( infoI != records.end() ){
```
Comment (tremblerz) [inline nmap_dns.cc]: This condition show our unconfidence in presence of record in map, I think we can remove this. I tested without this condition also and it works fine.

**Comment by dmiller-nmap** [comment]
@tremblerz This appears to have solved the earlier crash, but now it crashes when trying to do the deferred list. Here are some example crashes:

```
Completed Parallel DNS resolution of 4096 hosts. at 13:13, 81.08s elapsed
Performing system-dns for 1 domain names that use CNAMEs
Initiating System CNAME DNS resolution of 1 host. at 13:13
nmap: Target.cc:275: int Target::TargetSockAddr(sockaddr_storage*, size_t*) const: Assertion `targetsocklen <= sizeof(*ss)' failed.
Aborted
```

and only specifically checking the problem address:

```
mass_rdns: CNAME found for <195.57.153.175>
CAPACITY <72.14.191.5> = 12
NSOCK INFO [0.2910s] nsock_iod_delete(): nsock_iod_delete (IOD #1)
NSOCK INFO [0.2910s] nevent_delete(): nevent_delete on event #66 (type READ)
NSOCK DEBUG [0.2910s] event_delete(): event_delete (IOD #1) (EID #66)
NSOCK INFO [0.2910s] nsock_iod_delete(): nsock_iod_delete (IOD #2)
NSOCK INFO [0.2910s] nevent_delete(): nevent_delete on event #34 (type READ)
NSOCK DEBUG [0.2910s] event_delete(): event_delete (IOD #2) (EID #34)
NSOCK INFO [0.2910s] nsock_iod_delete(): nsock_iod_delete (IOD #3)
NSOCK INFO [0.2910s] nevent_delete(): nevent_delete on event #50 (type READ)
NSOCK DEBUG [0.2910s] event_delete(): event_delete (IOD #3) (EID #50)
CAPACITY <72.14.191.5> = 14
mass_rdns: OK MATCHED <195.57.153.175> to <175.red-195-57-153.customer.static.ccgg.telefonica.net>
mass_rdns: 0.29s 0/1 [#: 3, OK: 0, NX: 0, DR: 0, SF: 0, TR: 1]
NSOCK DEBUG [0.2910s] event_delete(): event_delete (IOD #1) (EID #18)
Completed Parallel DNS resolution of 1 host. at 14:54, 0.29s elapsed
Performing system-dns for 1 domain names that use CNAMEs
Initiating System CNAME DNS resolution of 1 host. at 14:54
Failed to get target socket address.
QUITTING!
```

Interestingly, this address has two answers, a regular PTR and a CNAME. We should add code in read_evt_handler to only check the CNAME if there's no PTR, similar to how we are trying to do for truncated in #434. This could be why we're getting memory corruption: we're handling 2 or more answers for a single question.

```
$ dig -x 195.57.153.175

; <<>> DiG 9.9.5-3ubuntu0.8-Ubuntu <<>> -x 195.57.153.175
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 45717
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 2, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;175.153.57.195.in-addr.arpa.   IN      PTR

;; ANSWER SECTION:
175.153.57.195.in-addr.arpa. 86400 IN   CNAME   175.0.153.57.195.in-addr.arpa.
175.0.153.57.195.in-addr.arpa. 86400 IN PTR     175.red-195-57-153.customer.static.ccgg.telefonica.net.

;; AUTHORITY SECTION:
0.153.57.195.in-addr.arpa. 86400 IN     NS      artemis.ttd.net.
0.153.57.195.in-addr.arpa. 86400 IN     NS      minerva.ttd.net.

;; Query time: 325 msec
;; SERVER: 72.14.179.5#53(72.14.179.5)
;; WHEN: Tue Jun 21 14:56:41 UTC 2016
;; MSG SIZE  rcvd: 192
```
```

**Comment by tremblerz** [comment]
I have considered previous implementation where we are again sending CNAME request though, unnecessarily.

**Pull request closed without merging**